x-pack/test/detection_engine_api_integration/security_and_spaces/rule_execution_logic/non_ecs_fields.ts

```diff
@@ -0,0 +1,101 @@
+
+// eslint-disable-next-line import/no-default-export
+export default ({ getService }: FtrProviderContext) => {
+  const supertest = getService('supertest');
+  const esArchiver = getService('esArchiver');
+  const es = getService('es');
+  const log = getService('log');
+
+  describe('Non ECS fields in alert document source', () => {
+    before(async () => {
+      await esArchiver.load(
+        'x-pack/test/functional/es_archives/security_solution/ecs_non_compliant'
+      );
+    });
+
+    after(async () => {
+      await esArchiver.unload(
+        'x-pack/test/functional/es_archives/security_solution/ecs_non_compliant'
+      );
+      await deleteSignalsIndex(supertest, log);
+      await deleteAllAlerts(supertest, log);
+    });
+
+    // source agent.name is object, ECS mapping for agent.name is keyword
+    it('should remove source object field from alert if ECS field mapping is keyword', async () => {
+      const rule: QueryRuleCreateProps = {
+        ...getRuleForSignalTesting(['ecs_non_compliant']),
+        query: `id: "agent.name is object"`,
+      };
+      const { previewId } = await previewRule({ supertest, rule });
+      const previewAlerts = await getPreviewAlerts({ es, previewId });
+
+      // valid ECS field is not getting removed
+      expect(previewAlerts[0]._source).toHaveProperty('agent.version', 'test-1');
+
+      // invalid ECS field 'agent.name' is getting removed
+      expect(previewAlerts[0]._source).not.toHaveProperty('agent.name');
+    });
+
+    // source container.image is keyword, ECS mapping for container.image is object
+    it('should remove source keyword field from alert if ECS field mapping is object', async () => {
+      const rule: QueryRuleCreateProps = {
+        ...getRuleForSignalTesting(['ecs_non_compliant']),
+        query: `id: "container.image is keyword"`,
+      };
+      const { previewId } = await previewRule({ supertest, rule });
+      const previewAlerts = await getPreviewAlerts({ es, previewId });
+
+      // invalid ECS field 'container.image' is getting removed
+      expect(previewAlerts[0]._source).not.toHaveProperty('container.image');
+    });
+
+    // source agent.type is long, ECS mapping for agent.type is keyword
+    it('should not remove source long field from alert if ECS field mapping is keyword', async () => {
+      const rule: QueryRuleCreateProps = {
+        ...getRuleForSignalTesting(['ecs_non_compliant']),
+        query: `id: "agent.type is long"`,
+      };
+      const { previewId, logs } = await previewRule({ supertest, rule });
+      const previewAlerts = await getPreviewAlerts({ es, previewId });
+
+      expect(logs[0].errors).toEqual([]);
+
+      // long value should be indexed as keyword
+      expect(previewAlerts[0]._source).toHaveProperty('agent.type', 13);
+    });
+
+    // source client.ip is keyword, ECS mapping for client.ip is ip
+    it('should remove source non ip field from alert if ECS field mapping is ip', async () => {
```

Comment:
@marshallmain, looks like we would need to validate manually each ES data type.
This test, currently failing with the following error:
```
 +   "Bulk Indexing of signals failed: failed to parse field [client.ip] of type [ip] in document with id 'ea4c710cdac74943906da99f62cd735f37e08469474641245c3dab95453044a2'. Preview of field's value: 'non-valid-ip',failed to parse field [client.ip] of type [ip] in document with id 'b345162e0f331f5de0a84c3d38cd8e9df908074662d78d2198e9f20f5b35150a'. Preview of field's value: 'non-valid-ip'",
```

It's because, source `client.ip` is of `keyword` type has value *non-valid-ip*, which is not valid ip data type.

There are quite few complex data types that exists: https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-types.html, like histogram, geo_point, range. If we want to make sure source data will be successfully saved in alerts index, we will need to check if source data against these types as well.
What do you think about this approach?

Reply:
after looking further in ECS, it turned out, not all ES types are present there.
So, only few of complex types are needed to be checked
- date
- ip
- geo_point

Reply:
I think if we can reliably validate these types then it would be nice to add, but I see it being tricky to account for all the different acceptable representations of a type. If we add this validation and don't account for all the possibilities then we risk removing fields that are actually valid. If we don't add this validation then as this test shows we still have the possibility of including fields that cause errors when writing alerts.

My preference would be to only add the validation for complex types where we can be confident that we won't remove fields that are actually valid, and for the other types continue to accept any (non-object) value and accept the risk of invalid values causing alerts to fail at write time. The alternative, potentially removing fields that are actually valid, feels like a bigger and riskier change that might cause some pain for users - whereas an incremental improvement is something we can continue to build on.

So for `date`, `ip`, and `geo_point`, I guess it's a question of how confident are we that we can write a validation function for each of those types that accepts all possible valid values for that type?